File: app/src/main/java/com/unsplash/client/network/response/ApiError.java
package com.unsplash.client.network.response;

import androidx.annotation.NonNull;

import com.google.gson.annotations.SerializedName;
import com.google.gson.internal.LinkedTreeMap;

public class ApiError extends Throwable {

    @SerializedName("code")
    private int code;

    @SerializedName("message")
    private String description;

    public int getCode() {
        return code;
    }

    public String getDescription() {
        return description;
    }

    @NonNull
    static ApiError fromErrorObject(Object error) {

        ApiError result = new ApiError();

        LinkedTreeMap map = (LinkedTreeMap) error;

        result.code = (int)((double) map.get("code"));
        result.description = (String) map.get("message");

        return result;
    }
}

File: app/src/main/java/com/unsplash/client/network/response/ApiError.kt
package com.unsplash.client.network.response

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import com.google.gson.internal.LinkedTreeMap

class ApiError : Throwable() {

    @SerializedName("code")
    @Expose
    var code: Int = 0

    @SerializedName("message")
    @Expose
    var description: String = ""

    companion object {

        @Suppress("UNCHECKED_CAST")
        fun fromErrorObject(error: Any): ApiError {

            val result = ApiError()

            val map: LinkedTreeMap<String, *> = error as LinkedTreeMap<String, *>

            result.code = (map["code"] as Double).toInt()
            result.description = map["message"] as String

            return result
        }
    }
}
